File: js/src/collection.js
/**
 * Collection envelope, in-memory query execution and RFC 8288 link headers
 * (specification §6).
 */

/**
 * Test whether an item satisfies every filter.
 *
 * A filter whose value is a list matches when the item field equals any member,
 * which is what repeated query parameters (`?tag=a&tag=b`) mean.
 *
 * @param {object} item - Candidate item
 * @param {object} filters - Field filters
 * @returns {boolean} True when the item matches
 */
export function matchesFilters(item, filters) {
  return Object.entries(filters).every(([field, expected]) => {
    const actual = item?.[field];
    if (Array.isArray(expected)) {
      return expected.some((candidate) => candidate === actual);
    }
    return actual === expected;
  });
}

/**
 * Compare two values in a total order that works across LINO scalar types.
 *
 * @param {*} left - First value
 * @param {*} right - Second value
 * @returns {number} Negative, zero or positive
 */
function compareValues(left, right) {
  if (left === right) {
    return 0;
  }
  if (left === undefined || left === null) {
    return -1;
  }
  if (right === undefined || right === null) {
    return 1;
  }
  if (typeof left === "number" && typeof right === "number") {
    return left - right;
  }
  return String(left).localeCompare(String(right));
}

/**
 * Sort items by the sort keys of a collection query.
 *
 * @param {object[]} items - Items to sort (not mutated)
 * @param {Array<{field: string, descending: boolean}>} sort - Sort keys
 * @returns {object[]} Sorted copy
 */
export function sortItems(items, sort) {
  if (!sort || sort.length === 0) {
    return items.slice();
  }
  return items.slice().sort((left, right) => {
    for (const { field, descending } of sort) {
      const comparison = compareValues(left?.[field], right?.[field]);
      if (comparison !== 0) {
        return descending ? -comparison : comparison;
      }
    }
    return 0;
  });
}

/**
 * Reduce an item to a sparse fieldset.
 *
 * @param {object} item - Item to project
 * @param {string[]|null} fields - Field names, or null for every field
 * @returns {object} Projected item
 */
export function projectFields(item, fields) {
  if (!fields) {
    return item;
  }
  const projected = {};
  for (const field of fields) {
    if (item && Object.hasOwn(item, field)) {
      projected[field] = item[field];
    }
  }
  return projected;
}

/**
 * Run a parsed collection query against an in-memory list.
 *
 * @param {object[]} items - Every item of the collection
 * @param {object} query - Query from {@link parseCollectionQuery}
 * @returns {{items: object[], page: {limit: number, offset: number, total: number, count: number}}} Collection envelope
 */
export function applyCollectionQuery(items, query) {
  const filtered = items.filter((item) => matchesFilters(item, query.filters));
  const sorted = sortItems(filtered, query.sort);
  const page = sorted.slice(query.offset, query.offset + query.limit);
  const projected = page.map((item) => projectFields(item, query.fields));

  return collectionEnvelope(projected, {
    limit: query.limit,
    offset: query.offset,
    total: filtered.length,
  });
}

/**
 * Wrap items in the collection envelope of the specification.
 *
 * @param {object[]} items - Items of this page
 * @param {{limit: number, offset: number, total: number}} page - Page metadata
 * @returns {{items: object[], page: object}} Collection envelope
 */
export function collectionEnvelope(items, page) {
  return {
    items,
    page: {
      limit: page.limit,
      offset: page.offset,
      total: page.total,
      count: items.length,
    },
  };
}

/**
 * Build the RFC 8288 `Link` header value for a paginated collection.
 *
 * @param {string} path - Request path without a query string
 * @param {object} query - Original query parameters
 * @param {{limit: number, offset: number, total: number}} page - Page metadata
 * @returns {string} `Link` header value ("" when there is nothing to link to)
 */
export function paginationLinkHeader(path, query, page) {
  const { limit, offset, total } = page;
  if (!limit) {
    return "";
  }

  const build = (targetOffset) => {
    const parameters = new URLSearchParams();
    for (const [name, value] of Object.entries(query ?? {})) {
      if (name === "limit" || name === "offset") {
        continue;
      }
      for (const item of Array.isArray(value) ? value : [value]) {
        parameters.append(name, String(item));
      }
    }
    parameters.set("limit", String(limit));
    parameters.set("offset", String(targetOffset));
    return `${path}?${parameters.toString()}`;
  };

  const lastOffset = total === 0 ? 0 : Math.floor((total - 1) / limit) * limit;
  const links = [];

  links.push(`<${build(0)}>; rel="first"`);
  if (offset > 0) {
    links.push(`<${build(Math.max(0, offset - limit))}>; rel="prev"`);
  }
  if (offset + limit < total) {
    links.push(`<${build(offset + limit)}>; rel="next"`);
  }
  links.push(`<${build(lastOffset)}>; rel="last"`);

  return links.join(", ");
}

File: rust/src/collection.rs
//! Collection envelope, in-memory query execution and RFC 8288 link headers
//! (specification section 6).

use lino_objects_codec::LinoValue;

use crate::query::{CollectionQuery, Filter, QueryParams, SortKey, form_encode};
use crate::value::{array, compare, get, int, object};

/// Test whether an item satisfies every filter.
///
/// A filter carrying more than one value matches when the item field equals any
/// of them, which is what repeated query parameters (`?tag=a&tag=b`) mean.
pub fn matches_filters(item: &LinoValue, filters: &[Filter]) -> bool {
    filters.iter().all(|filter| {
        let actual = get(item, &filter.field).cloned().unwrap_or(LinoValue::Null);
        filter.values.contains(&actual)
    })
}

/// Sort items by the sort keys of a collection query.
///
/// The original list is left untouched, and the sort is stable, so items that
/// compare equal keep the order the store returned them in.
pub fn sort_items(items: &[LinoValue], sort: &[SortKey]) -> Vec<LinoValue> {
    let mut sorted = items.to_vec();
    if sort.is_empty() {
        return sorted;
    }
    sorted.sort_by(|left, right| {
        for key in sort {
            let left_value = get(left, &key.field).cloned().unwrap_or(LinoValue::Null);
            let right_value = get(right, &key.field).cloned().unwrap_or(LinoValue::Null);
            let comparison = compare(&left_value, &right_value);
            if comparison != std::cmp::Ordering::Equal {
                return if key.descending {
                    comparison.reverse()
                } else {
                    comparison
                };
            }
        }
        std::cmp::Ordering::Equal
    });
    sorted
}

/// Reduce an item to a sparse fieldset.
pub fn project_fields(item: &LinoValue, fields: Option<&Vec<String>>) -> LinoValue {
    let Some(fields) = fields else {
        return item.clone();
    };
    let Some(members) = crate::value::as_object(item) else {
        return item.clone();
    };
    LinoValue::Object(
        fields
            .iter()
            .filter_map(|field| {
                members
                    .iter()
                    .find(|(name, _)| name == field)
                    .map(|(name, value)| (name.clone(), value.clone()))
            })
            .collect(),
    )
}

/// Wrap items in the collection envelope of the specification.
///
/// # Examples
///
/// ```
/// use lino_rest_api::collection::collection_envelope;
/// use lino_rest_api::value::{get, int};
///
/// let envelope = collection_envelope(vec![int(1)], 20, 0, 1);
/// assert_eq!(get(get(&envelope, "page").unwrap(), "count"), Some(&int(1)));
/// ```
pub fn collection_envelope(
    items: Vec<LinoValue>,
    limit: usize,
    offset: usize,
    total: usize,
) -> LinoValue {
    let count = items.len();
    object([
        ("items", array(items)),
        (
            "page",
            object([
                ("limit", int(limit as i64)),
                ("offset", int(offset as i64)),
                ("total", int(total as i64)),
                ("count", int(count as i64)),
            ]),
        ),
    ])
}

/// Run a parsed collection query against an in-memory list.
pub fn apply_collection_query(items: &[LinoValue], query: &CollectionQuery) -> LinoValue {
    let filtered: Vec<LinoValue> = items
        .iter()
        .filter(|item| matches_filters(item, &query.filters))
        .cloned()
        .collect();
    let ordered = sort_items(&filtered, &query.sort);
    let page: Vec<LinoValue> = ordered
        .into_iter()
        .skip(query.offset)
        .take(query.limit)
        .map(|item| project_fields(&item, query.fields.as_ref()))
        .collect();

    collection_envelope(page, query.limit, query.offset, filtered.len())
}

/// Build the RFC 8288 `Link` header value for a paginated collection.
///
/// The value is empty when the page size is zero, because there is then nothing
/// to page through.
pub fn pagination_link_header(
    path: &str,
    query: &QueryParams,
    limit: usize,
    offset: usize,
    total: usize,
) -> String {
    if limit == 0 {
        return String::new();
    }

    let build = |target_offset: usize| {
        let mut parameters: Vec<String> = query
            .iter()
            .filter(|(name, _)| *name != "limit" && *name != "offset")
            .map(|(name, value)| format!("{}={}", form_encode(name), form_encode(value)))
            .collect();
        parameters.push(format!("limit={limit}"));
        parameters.push(format!("offset={target_offset}"));
        format!("{path}?{}", parameters.join("&"))
    };

    let last_offset = if total == 0 {
        0
    } else {
        ((total - 1) / limit) * limit
    };
    let mut links = vec![format!("<{}>; rel=\"first\"", build(0))];
    if offset > 0 {
        links.push(format!(
            "<{}>; rel=\"prev\"",
            build(offset.saturating_sub(limit))
        ));
    }
    if offset + limit < total {
        links.push(format!("<{}>; rel=\"next\"", build(offset + limit)));
    }
    links.push(format!("<{}>; rel=\"last\"", build(last_offset)));

    links.join(", ")
}
